### PR TITLE
WIP: Potential optimizations

### PR DESCRIPTION
**Pull request by RuslanZavacky**
I am trying to find ways if possible to improve performance in our monorepo, and debugging running ember s got me to the fs-updater, merge-trees and broccoli-merge-trees.

Here are some data for the context:

<details>
  <summary>Show screenshot of debug and functions time</summary>
Screenshot 2022-01-26 at 12 10 18
Screenshot 2022-01-26 at 12 09 15
</details>

From these screenshots, couple things that are interesting:

1. A lot of time spent in `mergeDirectories` (hence my investigation)
2. 30% of time spent in GC - which is due to the size of project and a lot of recursion to read directories

This change that I've identified should potentially help minimize `fs.stat` calls, and on our monorepo, it does seem like it is a lot, here are some numbers:

```
  { lstat: 159039, dirent: 363220 }
```

`dirent` number shows how many times instead of fs.stat we were able to derive information required to determine directory/file/symbolic link from the additional info from the `fs.readdir`. This is particularly important on the systems with software that is actively monitor/backup all the changes (in enterprises).

I'd greatly appreciate any feedback/hints/help to figure out if there is more that I can look at to get much better results and optimizations in general.

I know there was this effort https://github.com/broccolijs/broccoli-merge-trees/pull/43 which potentially can lead to quite a good improvement on bigger projects as well.

Reference to other PRs that might improve performance in general:
https://github.com/broccolijs/node-merge-trees/pull/6
https://github.com/broccolijs/broccoli-persistent-filter/pull/188
https://github.com/broccolijs/fs-merger

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -226,7 +298,7 @@ function fileStatsEqual(a, b) {
   // Note that stats.mtimeMs is only available as of Node 9
   return (
     a.ino === b.ino &&
-    a.mtime.getTime() === b.mtime.getTime() &&
+    // a.mtime.getTime() === b.mtime.getTime() &&
```
Comment (NullVoxPopuli) [inline index.js:301]: is/was this important?

File: index.js [hunk 2]
```diff
@@ -11,25 +11,47 @@ let loggerGen = require("heimdalljs-logger");
...
 
   getIndexSync() {
-    if (this._index != null) return this._index;
+    if (this._index != null) {
+      if (this._isNodeModulesPackage) {
+        return this._index;
+      }
+
+      const now = Date.now();
+      const isCacheValid = (this._cacheBuildAt + CACHE_TTL) > now;
```
Comment (NullVoxPopuli) [inline index.js:37]: how does this affect the dev-time reload of linked packages that are maybe `ctrl+s`'d rapidly?
Like, does this mean that when working on a library, we can only receive updates at quickest, every 5s?

File: index.js [hunk 3]
```diff
@@ -64,15 +90,61 @@ function makeFSObject(p) {
...
+
+    if (dirent.isFile()) {
+      return getFileClass(p);
+    }
+
+    if (!dirent.isSymbolicLink()) {
+      throw new Error("File has unexpected type: " + p);
+    }
+  } else {
+    let stats = fs.lstatSync(p);
```
Comment (NullVoxPopuli) [inline index.js:131]: any difference in perf in sync vs async?

Comment (lifeart) [reply index.js:131]: likely yes (for SSD drives, because there is  more file threads available), but this package is sync

File: index.js (hunk 1 above)
Comment (lifeart) [reply index.js:301]: @NullVoxPopuli in real life it's really rare case where file is changed without it's changed size

File: index.js (hunk 3 above)
Comment (lifeart) [reply index.js:131]: but for stat it's not clear - https://github.com/nodejs/node/issues/38006

File: index.js (hunk 2 above)
Comment (RuslanZavacky) [reply index.js:37]: For now yes, it was more of an experiment and solution to our problem.. with this - we have issues in tests for the merge-trees, so need to figure out some other solution.. but to say, it does add a big speed boost, so caching is a must.. need to think how to pass it probably from the consuming packages.

**Pull request closed without merging**